substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1033,18 +1073,36 @@ impl<T: Config> Pallet<T> {
 	/// wrong.
 	pub fn do_add_validator(who: &T::AccountId, prefs: ValidatorPrefs) {
 		if !Validators::<T>::contains_key(who) {
-			// maybe update sorted list.
-			let _ = T::VoterList::on_insert(who.clone(), Self::weight_of(who))
-				.defensive_unwrap_or_default();
-		}
+			let self_stake = Self::stake(who);
+			T::EventListeners::on_validator_add(
+				who,
+				Some(self_stake.defensive_unwrap_or_default().into()),
+			);
+		};
 		Validators::<T>::insert(who, prefs);
 
 		debug_assert_eq!(
 			Nominators::<T>::count() + Validators::<T>::count(),
-			T::VoterList::count()
+			T::VoterList::iter()
+				.filter(|v| Self::status(&v) != Ok(StakerStatus::Idle))
+				.count() as u32,
 		);
 	}
 
+	/// Tries to chill a validator.
+	///
+	/// A chilled validator is removed from the `Validators` map.
+	pub(crate) fn do_chill_validator(who: &T::AccountId) -> bool {
+		match Validators::<T>::contains_key(who) {
```

Comment:
nit: would probably look better with
```
if condition {
   ...
   return true;
}
false
```
or you could use the new `bool` mechanic
```
Validators::<T>::contains_key(who)
   .then(|| {
      Validators::<T>::remove(who);
      T::EventListeners::on_validator_idle(who);
   })
   .is_some()
```